### PR TITLE
PT-3751 full text search

### PR DESCRIPTION
**Pull request by alexbukata**

### CODE REVIEW

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt [hunk 1]
```diff
@@ -35,6 +35,7 @@ class EsItemOptimizedSearchService(
         }
         val max = PageSize.ITEM.limit(limit)
         return when (sort.type) {
+            EsItemSortType.MOST_RELEVANT,
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:38]: there is no sense in optimizing MOST_RELEVANT. it should go just above TRAIT

File: core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsItemSort.kt [hunk 2]
```diff
@@ -14,7 +14,7 @@ data class EsItemSort(
         }
 
     companion object {
-        val DEFAULT = EsItemSort(type = EsItemSortType.DEFAULT)
+        val DEFAULT = EsItemSort(type = EsItemSortType.MOST_RELEVANT)
```
Comment (dmitry-saraykin) [inline core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsItemSort.kt:17]: should be LATEST_FIRST

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemElasticService.kt [hunk 3]
```diff
@@ -376,6 +376,7 @@ class ItemElasticService(
     private fun convertSort(request: ItemsSearchRequestDto): EsItemSort {
         val sort = request.sort ?: return EsItemSort.DEFAULT
         return when (sort) {
+            ItemsSearchSortDto.DEFAULT -> EsItemSort(type = EsItemSortType.MOST_RELEVANT)
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemElasticService.kt:379]: Why DEFAULT? Let's call it RELEVANCE

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt [hunk 4]
```diff
@@ -54,17 +56,24 @@ class ItemFilterConverter() {
     }
 
     fun searchItems(filter: ItemsSearchFilterDto, cursor: String?): EsItemFilter {
+        filter.fullText
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:59]: remove?

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt [hunk 5]
```diff
@@ -54,17 +56,24 @@ class ItemFilterConverter() {
...
             creators = filter.creators?.map { it.fullId() }?.toSet().orEmpty(),
             mintedFrom = filter.mintedAtFrom,
             mintedTo = filter.mintedAtTo,
             updatedFrom = filter.lastUpdatedAtFrom,
             updatedTo = filter.lastUpdatedAtTo,
             deleted = filter.deleted,
+            fullText = filter.fullText?.let {
+                FullTextSearch(
+                    it.text,
+                    it.fields?.map { tf -> TextField.valueOf(tf.name.uppercase()) } ?: listOf(TextField.NAME)
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:73]: better not to use valueOf as it is a runtime exception. Better to use `when` which will give compile time exception

Comment (dmitry-saraykin) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:73]: also please add arg names

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt [hunk 6]
```diff
@@ -54,17 +56,24 @@ class ItemFilterConverter() {
...
+                )
+            },
+            names = filter.names?.toSet(),
```
Comment (dmitry-saraykin) [inline api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:76]: should we remove names and descriptions as they are not working at the moment?

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt (hunk 1 above)
Comment (alexbukata) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/EsItemOptimizedSearchService.kt:38]: done

File: core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsItemSort.kt (hunk 2 above)
Comment (alexbukata) [reply core/src/main/kotlin/com/rarible/protocol/union/core/model/elastic/EsItemSort.kt:17]: done

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemElasticService.kt (hunk 3 above)
Comment (alexbukata) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemElasticService.kt:379]: done

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt (hunk 4 above)
Comment (alexbukata) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:59]: done

File: api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt (hunk 5 above)
Comment (alexbukata) [reply api/src/main/kotlin/com/rarible/protocol/union/api/service/elastic/ItemFilterConverter.kt:73]: done

**Review by dmitry-saraykin — APPROVED** [review]

**Pull request merged**